### PR TITLE
RBAC support for downloading scripts with managed identities

### PR DESCRIPTION
**Pull request by D1v38om83r**
Enable CSE linux to use managed identities to access azure storage blobs to download scripts.

### CODE REVIEW

File: pkg/download/blobwithmsitoken.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	return &blobWithMsiToken{url, msiProvider}
+}
+
+func GetMsiProviderForStorageAccountsImplicitly(blobUri string) MsiProvider {
+	msiProvider := msi.NewMsiProvider(httputil.NewSecureHttpClient(httputil.DefaultRetryBehavior))
+	return func() (msi.Msi, error) { return msiProvider.GetMsiForResource(GetResourceNameFromBlobUri(blobUri)) }
+}
+
+func GetMsiProviderForStorageAccountsWithClientId(blobUri, clientId string) MsiProvider {
+	msiProvider := msi.NewMsiProvider(httputil.NewSecureHttpClient(httputil.DefaultRetryBehavior))
```
Comment (vivlingaiah) [inline pkg/download/blobwithmsitoken.go]: should we check clientId is valid here ?

File: pkg/download/blobwithmsitoken.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	return func() (msi.Msi, error) {
+		return msiProvider.GetMsiUsingClientId(clientId, GetResourceNameFromBlobUri(blobUri))
+	}
+}
+
+func GetMsiProviderForStorageAccountsWithObjectId(blobUri, objectId string) MsiProvider {
+	msiProvider := msi.NewMsiProvider(httputil.NewSecureHttpClient(httputil.DefaultRetryBehavior))
+	return func() (msi.Msi, error) {
```
Comment (vivlingaiah) [inline pkg/download/blobwithmsitoken.go]: should we check objectid is valid here ?

File: main/handlersettings.go [hunk 3]
```diff
@@ -66,6 +68,16 @@ func (h handlerSettings) validate() error {
 		return errStoragePartialCredentials
 	}
 
+	if (h.protectedSettings.StorageAccountKey != "" || h.protectedSettings.StorageAccountName != "") && !h.protectedSettings.ManagedIdentity.isEmpty() {
```
Comment (vivlingaiah) [inline main/handlersettings.go]: do we need a null check for h.protectedSettings.ManagedIdentity ?

File: main/schema.go [hunk 4]
```diff
@@ -73,7 +73,23 @@ const (
       "description": "Key for the Azure Storage Account (a base64 encoded string)",
       "type": "string",
       "pattern": "^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{4})$"
-    }
+    },
+	"managedIdentity": {
```
Comment (vivlingaiah) [inline main/schema.go]: correct indent

File: pkg/download/blobwithmsitoken.go (hunk 1 above)
Comment (D1v38om83r) [reply pkg/download/blobwithmsitoken.go]: Already checked as a part of json schema check https://github.com/Azure/custom-script-extension-linux/blob/10710fb27d7fb5061d9193f3f6edbb8b82b1a5a8/main/schema.go#L89

File: pkg/download/blobwithmsitoken.go (hunk 2 above)
Comment (D1v38om83r) [reply pkg/download/blobwithmsitoken.go]: Already checked as a part of JSON schema check

https://github.com/Azure/custom-script-extension-linux/blob/10710fb27d7fb5061d9193f3f6edbb8b82b1a5a8/main/schema.go#L84

File: main/handlersettings.go (hunk 3 above)
Comment (D1v38om83r) [reply main/handlersettings.go]: Good catch, that was definitely a bug

File: pkg/download/blobwithmsitoken.go (hunk 1 above)
Comment (vivlingaiah) [reply pkg/download/blobwithmsitoken.go]: I meant whether clientId has permissions

File: main/schema.go (hunk 4 above)
Comment (D1v38om83r) [reply main/schema.go]: fixed in new commit

File: pkg/download/blobwithmsitoken.go (hunk 1 above)
Comment (D1v38om83r) [reply pkg/download/blobwithmsitoken.go]: The only way to figure that is to actually attempt the download. If it doesn't have permissions, the download will fail and the set-extensions call itself with fail

File: pkg/download/blobwithmsitoken.go [hunk 5]
```diff
@@ -57,20 +57,38 @@ func NewBlobWithMsiDownload(url string, msiProvider MsiProvider) Downloader {
 
 func GetMsiProviderForStorageAccountsImplicitly(blobUri string) MsiProvider {
 	msiProvider := msi.NewMsiProvider(httputil.NewSecureHttpClient(httputil.DefaultRetryBehavior))
-	return func() (msi.Msi, error) { return msiProvider.GetMsiForResource(GetResourceNameFromBlobUri(blobUri)) }
+	return func() (msi.Msi, error) {
+		msi, err := msiProvider.GetMsiForResource(GetResourceNameFromBlobUri(blobUri))
+		if err != nil {
+			return msi, errors.Wrapf(err, "Unable to get managed identity. "+
+				"Please make sure that system assigned managed identity is enabled on the VM"+
```
Comment (vivlingaiah) [inline pkg/download/blobwithmsitoken.go]: add space in end "enabled on the VM "

File: pkg/download/downloader.go [hunk 6]
```diff
@@ -52,5 +52,13 @@ func Download(d Downloader) (int, io.ReadCloser, error) {
 	if resp.StatusCode == http.StatusOK {
 		return resp.StatusCode, resp.Body, nil
 	}
-	return resp.StatusCode, nil, fmt.Errorf("unexpected status code: got=%d expected=%d", resp.StatusCode, http.StatusOK)
+
+	err = fmt.Errorf("unexpected status code: got=%d expected=%d", resp.StatusCode, http.StatusOK)
```
Comment (vivlingaiah) [inline pkg/download/downloader.go]: use "actual" instead of "got"

File: pkg/download/downloader.go [hunk 7]
```diff
@@ -52,5 +52,13 @@ func Download(d Downloader) (int, io.ReadCloser, error) {
...
+	switch d.(type) {
+	case *blobWithMsiToken:
+		if resp.StatusCode == http.StatusForbidden {
```
Comment (vivlingaiah) [inline pkg/download/downloader.go]: do you get unauthorized (http 401) in any case ?

File: pkg/download/blobwithmsitoken.go (hunk 5 above)
Comment (D1v38om83r) [reply pkg/download/blobwithmsitoken.go]: fixed

File: pkg/download/downloader.go (hunk 7 above)
Comment (D1v38om83r) [reply pkg/download/downloader.go]: We only get 404's

File: pkg/download/downloader.go (hunk 6 above)
Comment (D1v38om83r) [reply pkg/download/downloader.go]: fixed

File: pkg/download/blobwithmsitoken.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+)
+
+const (
+	xMsVersionHeaderName = "x-ms-version"
+	xMsVersionValue      = "2018-03-28"
+	storageResourceName  = "https://storage.azure.com/"
+)
+
+var azureBlobDomains = map[string]interface{}{ // golang doesn't have builtin hash sets, so this is a workaround for that
+	"blob.core.windows.net":       nil,
```
Comment (vivlingaiah) [inline pkg/download/blobwithmsitoken.go]: shouldn't it be true instead of nil ?

Comment (D1v38om83r) [reply pkg/download/blobwithmsitoken.go]: Its inconsequential, I am only using the presence/absence of the key in the map/dictionary. I am not using the value.

**Pull request merged**